### PR TITLE
Add option to Operator to disable the reconciler for Services

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

Adds a new CLI flag to the Dapr Operator service `--disable-service-reconciler` to disable the built-in reconciler (default=false - i.e. the reconciler is enabled)

The Operator has multiple roles in Dapr, and one of them is having a reconciler that creates Services for each Dapr-enabled Deployment and StatefulSet. In some cases, this reconciler may conflict with other reconcilers that are running on the cluster. Additionally, it uses a lot of memory, even in clusters that do not need it.

Also added an option to the Helm chart to control this flag.

### CODE REVIEW

File: pkg/operator/handlers/dapr_handler.go [hunk 1]
```diff
@@ -16,17 +15,14 @@ import (
...
 
+	"github.com/dapr/dapr/pkg/injector/annotations"
 	"github.com/dapr/dapr/pkg/operator/monitoring"
 	"github.com/dapr/dapr/pkg/validation"
 	"github.com/dapr/dapr/utils"
 	"github.com/dapr/kit/logger"
 )
 
 const (
-	daprEnabledAnnotationKey        = "dapr.io/enabled"
```
Comment (ItalyPaleAle) [inline pkg/operator/handlers/dapr_handler.go:26]: Reviewers: these constants were already defined in pkg/injector/annotations so importing them from there instead

**Comment by yaron2** [comment]
It should be mentioned that enabling this will break/disable service to service invocation.

**Comment by ItalyPaleAle** [comment]
@yaron2 Updated the README to add the warning, thanks.

File: pkg/operator/handlers/dapr_handler.go [hunk 2]
```diff
@@ -212,23 +226,23 @@ func (h *DaprHandler) createDaprServiceValues(ctx context.Context, expectedServi
...
 
 	if enableMetrics {
-		annotations["prometheus.io/probe"] = "true"
-		annotations["prometheus.io/scrape"] = "true" // WARN: deprecated as of v1.7 please use prometheus.io/probe instead.
-		annotations["prometheus.io/port"] = strconv.Itoa(metricsPort)
-		annotations["prometheus.io/path"] = "/"
+		annotationsMap["prometheus.io/probe"] = "true"
+		annotationsMap["prometheus.io/scrape"] = "true" // WARN: deprecated as of v1.7 please use prometheus.io/probe instead.
+		annotationsMap["prometheus.io/port"] = strconv.Itoa(metricsPort)
+		annotationsMap["prometheus.io/path"] = "/"
```
Comment (mukundansundar) [inline pkg/operator/handlers/dapr_handler.go:237]: can these keys be defined as constants?

File: pkg/operator/handlers/dapr_handler.go [hunk 3]
```diff
@@ -264,37 +278,32 @@ func (h *DaprHandler) createDaprServiceValues(ctx context.Context, expectedServi
...
+	return annotationsMap[annotations.KeyAppID]
 }
 
 func (h *DaprHandler) isAnnotatedForDapr(wrapper ObjectWrapper) bool {
-	annotations := wrapper.GetTemplateAnnotations()
-	enabled, ok := annotations[daprEnabledAnnotationKey]
-	if !ok {
+	annotationsMap := wrapper.GetTemplateAnnotations()
+	val := annotationsMap[annotations.KeyEnabled]
+	if val == "" {
```
Comment (mukundansundar) [inline pkg/operator/handlers/dapr_handler.go:288]: why was this changed to not use ok?

File: pkg/operator/handlers/dapr_handler.go (hunk 2 above)
Comment (ItalyPaleAle) [reply pkg/operator/handlers/dapr_handler.go:237]: Done

File: pkg/operator/handlers/dapr_handler.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/operator/handlers/dapr_handler.go:292]: Good question.

First, context: `val, ok := annotationsMap[annotations.KeyEnabled]`:

- If the key does not exist on the map, `val == ""`
- You can have `val == ""` also if the key exists, but it's empty. That's where the `ok` is useful: in this case, `ok == true`

Now let's take a look at the behavior of the previous code vs current:

**Previous**
```go
enabled, ok := annotations[daprEnabledAnnotationKey]
if !ok {
	return false
}
return utils.IsTruthy(enabled)
```

**Current**

```go
val := annotationsMap[annotations.KeyEnabled]
if val == "" {
	return false
}
return utils.IsTruthy(val)
```

As you can see, in both cases the if the value doesn't exist, the returned value is `val == ""` so the function returns false. The additional check for `ok` was redundant.

File: charts/dapr/README.md [hunk 4]
```diff
@@ -126,6 +126,7 @@ The Helm chart has the follow configuration options that can be supplied:
 | `dapr_operator.runAsNonRoot`              | Boolean value for `securityContext.runAsNonRoot`. You may have to set this to `false` when running in Minikube | `true` |
 | `dapr_operator.resources`                 | Value of `resources` attribute. Can be used to set memory/cpu resources/limits. See the section "Resource configuration" above. Defaults to empty | `{}` |
 | `dapr_operator.debug.enabled`             | Boolean value for enabling debug mode | `{}` |
+| `dapr_operator.serviceReconciler`         | Boolean value for enabling the reconciler that creates Services for Dapr-enabled Deployments and StatefulSets.<br>Note: disabling this could prevent Dapr service invocation from working. | `true` |
```
Comment (mukundansundar) [inline charts/dapr/README.md:129]: should this key be `dapr_operator.serviceReconciler.enabled` ?
`dapr_operator.runAsNonRoot` implicitly means that it is a bool value. But the above does not.
Having enabled or disabled would be better here IMO...

File: charts/dapr/charts/dapr_operator/values.yaml [hunk 5]
```diff
@@ -16,6 +16,8 @@ fullnameOverride: ""
 
 runAsNonRoot: true
 
+serviceReconciler: true
```
Comment (mukundansundar) [inline charts/dapr/charts/dapr_operator/values.yaml:19]: Would prefer to have a disable flag here instead of an enable flag.
Because the default behavior is enabled. So providing a flag to disable it makes it more consistent.

File: cmd/operator/main.go [hunk 6]
```diff
@@ -114,8 +115,9 @@ func init() {
 
 	flag.StringVar(&watchInterval, "watch-interval", defaultWatchInterval, "Interval for polling pods' state, e.g. '2m'. Set to '0' to disable, or 'once' to only run once when the operator starts")
 	flag.IntVar(&maxPodRestartsPerMinute, "max-pod-restarts-per-minute", defaultMaxPodRestartsPerMinute, "Maximum number of pods in an invalid state that can be restarted per minute")
-	flag.BoolVar(&disableLeaderElection, "disable-leader-election", false, "Disable leader election for operator")
 
+	flag.BoolVar(&disableLeaderElection, "disable-leader-election", false, "Disable leader election for operator")
+	flag.BoolVar(&disableServiceReconciler, "disable-service-reconciler", false, "Disable the Service reconciler for Dapr-enabled Deployments and StatefulSets")
```
Comment (mukundansundar) [inline cmd/operator/main.go:120]: nitpick: if grouping can group Bool vars together.

File: pkg/operator/handlers/dapr_handler.go [hunk 7]
```diff
@@ -78,11 +79,13 @@ func (h *DaprHandler) Init() error {
 				return nil
 			}
 			return []string{owner.Name}
-		}); err != nil {
+		},
+	)
+	if err != nil {
```
Comment (mukundansundar) [inline pkg/operator/handlers/dapr_handler.go:84]: why is the short if changed to this format?

File: cmd/operator/main.go [hunk 8]
```diff
@@ -66,6 +66,7 @@ func main() {
 		WatchdogInterval:          0,
 		WatchdogMaxRestartsPerMin: maxPodRestartsPerMinute,
 		WatchNamespace:            watchNamespace,
+		ServiceReconcilerEnabled:  !disableServiceReconciler,
```
Comment (mukundansundar) [inline cmd/operator/main.go:69]: is it possible to add a unit test or e2e test for this?

File: pkg/operator/handlers/dapr_handler.go (hunk 7 above)
Comment (ItalyPaleAle) [reply pkg/operator/handlers/dapr_handler.go:84]: IMHO a function inside an if was unreadable :)

File: cmd/operator/main.go (hunk 8 above)
Comment (ItalyPaleAle) [reply cmd/operator/main.go:69]: It is not, actually.... It would require creating an ad-hoc Dapr cluster so it doesn't fit with our E2E test framework.

File: charts/dapr/README.md (hunk 4 above)
Comment (ItalyPaleAle) [reply charts/dapr/README.md:129]: Done (using `disabled` as per below)

File: charts/dapr/charts/dapr_operator/values.yaml (hunk 5 above)
Comment (ItalyPaleAle) [reply charts/dapr/charts/dapr_operator/values.yaml:19]: Done

File: cmd/operator/main.go (hunk 6 above)
Comment (ItalyPaleAle) [reply cmd/operator/main.go:120]: They're currently grouped by kind. These 3 are all about configuring the reconciler

**Comment by mukundansundar** [comment]
@ItalyPaleAle Can you change back to having the `enabled` config option as per helm [convention](https://helm.sh/docs/chart_best_practices/dependencies/ )

**Comment by ItalyPaleAle** [comment]
@mukundansundar done

**Review by mukundansundar — APPROVED** [review]
lgtm
